src/smt/theory_str_noodler/parikh_image.cpp

```diff
@@ -0,0 +1,449 @@
+            sum_symb.insert( {as, LenNode(LenFormulaType::PLUS)});
+        }
+
+        LenNode phi_cnt(LenFormulaType::AND);
+        for(const auto& [trans, var] : trans_vars) {
+            // set of atomic symbols
+            auto symb = this->ca.alph.get_symbol(std::get<1>(trans));
+            for(const ca::AtomicSymbol& as : symb) {
+                sum_symb.at(as).succ.emplace_back(LenNode(var));
+            }
+        }
+
+        // generate equations var_a = #t_1 + #t_2 ... where #t_1 is variable of a 
+        // transition containing in symbol set a
+        for (const auto& [as, var] : this->symbol_var) {
+            const LenNode& act = sum_symb.at(as);
+            if(act.succ.size() > 0) {
+                phi_cnt.succ.push_back(LenNode(LenFormulaType::EQ, {
+                    this->symbol_var.at(as),
+                    act
+                }));
+            }
+        }
+
+        return phi_cnt;
+    }
+
+    /**
+     * @brief Get mismatch formula for particular positions @p i and @p j. 
+     * For x_1 ... x_n != y_1 ... y_m create 
+     * mismatch(i,j) := #<L,x_1> + ... + #<L,x_{i-1}> + #<P,x_i,0> = #<L,y_1> + ... + #<L,y_{j-1}> + #<P,y_j,1> && #<P,x_i,0> >= 1 && #<P,y_j,1> >= 1
+     * where #symb is LIA variable counting number of occurrences of the symbol symb during 
+     * the run. Stored in this->symbol_var and computed by symbol_count_formula.
+     * 
+     * E.g., for x_1 x_2 x_3 != y_1 y_2 create a formula 
+     * mismatch(3,2) := #<L,x_1> + #<L,x_2> + #<P,x_3,0> = #<L,y_1> + #<P,y_2,1> && 
+     *      #<P,x_3,0> >= 1 && #<P,x_3,0> >= 1
+     * 
+     * 
+     * @param i Position on the left side of @p diseq
+     * @param j Position on the right side of @p diseq
+     * @param diseq Diseq
+     * @return LenNode mismatch(i,j)
+     */
+    LenNode ParikhImageCA::get_mismatch_formula(size_t i, size_t j, const Predicate& diseq) {
+        auto concat_len = [&](const Concat& con, int max_ind) -> LenNode {
+            LenNode sum_len(LenFormulaType::PLUS);
+            int ind = 0;
+            for(const BasicTerm& bt : con) {
+                if (ind > max_ind) {
+                    break;
+                }
+                ca::AtomicSymbol as = {0, bt, 0, 0}; // <L,x> symbol
+                sum_len.succ.push_back(LenNode(this->symbol_var.at(as)));
+                ++ind;
+            }
+            return sum_len;
+        };
+
+        LenNode left = concat_len(diseq.get_left_side(), i - 1);
+        // add symbol <P, var, 1, 0> where var is i-th variable on left side of diseq
+        ca::AtomicSymbol lats = {1, diseq.get_left_side()[i], 1, 0};
+        left.succ.push_back(LenNode(this->symbol_var.at(lats)));
+
+        LenNode right = concat_len(diseq.get_right_side(), j-1);
+        // add symbol <P, var, 2, 0> where var is j-th variable on left side of diseq
+        ca::AtomicSymbol rats = {1, diseq.get_right_side()[j], 2, 0};
+        right.succ.push_back(LenNode(this->symbol_var.at(rats)));
```

Comment:
The second paragraph is clear to me. Regarding the first one: You say that if x and y are the mismatch variables, I take the symbol `<P,x,1>` if `x > y`? Otherwise I take `<P,x,2>`?

Reply:
Great, thanks!